### ISSUE TITLE
The `git config -f`command doesn't work with a file of network folder for me.

### ISSUE THREAD

**Issue by Andrey-Bushman** [git-for-windows/git#241]
Git for Windows 2.4.5 x64.
It works fine with a local config-file:
`git config -f "d:\\dev\\settings\\gitconfig.txt" http.proxy http://@proxy2:8080`

I need to do the same with a config-file which is located at the network in our domain. This config-file is common for all developers of our company and has common settings (the proxy settings, for example). I have full rights for working with that directory. This is my directory only. It was created special for me by admin (an hour ago) and only I have full access to it.

I launch this:
`git config -f "\\\\hyprostr\\dfs\\groups\\developers\\settings\\gitconfig.txt" http.proxy http://@proxy2:8080`

But it doesn't work. I see the message:
![_0](https://cloud.githubusercontent.com/assets/10530134/8651865/2477260e-2984-11e5-9634-b7d4ab73e081.png)

if I press `y` then I get the same message again... I reloaded my computer, but it didn't help.

Through Windows Explorer I create, edit, remove and rename files and directories without problems in this location. 

Why does it happen and how can I fix it?

This works fine (from Git Bash): :
`echo 123 > "\\\\hyprostr\\dfs\\groups\\developers\\settings\\gitconfig.txt"`

and this works fine too:
`mv "\\\\hyprostr\\dfs\\groups\\developers\\settings\\gitconfig.txt" "\\\\hyprostr\\dfs\\groups\\developers\\settings\\gitconfig2.txt"`

So, I don't understand the reason of the problem... :(((

**Comment by Andrey-Bushman** [comment]
I see the `gitconfig.txt.lock` at the same directory. It was created by Git but not removed (I don't know why). I removed it manually and now it works fine.

**Comment by kostix** [comment]
Windows has no problems interpreting `/` instead of `\` which could make your life significantly easier:

```
kostix@programmer MINGW32 ~
$ git config --file //tigra/kostix/foo.conf --list
core.whatever=bar

kostix@programmer MINGW32 ~
$ cat //tigra/kostix/foo.conf
[core]
whatever = bar

kostix@programmer MINGW32 ~
$
```

I think the same applies to #239.

**Comment by Andrey-Bushman** [comment]
Oh.... Problem is not solved.

**Comment by Andrey-Bushman** [comment]
Look this video: http://www.youtube.com/watch?v=kudZ8jTg7Ww&feature=youtu.be

**Comment by Andrey-Bushman** [comment]
Is it a bug?

**Comment by dscho** [comment]
@Andrey-Bushman have you looked in detail at [@kostix's answer](https://github.com/git-for-windows/git/issues/241#issuecomment-120959612)? I believe that his advice essentially resolves this ticket.

**Comment by Andrey-Bushman** [comment]
No, using `/` instead of `\\` doesn't help. Pay attention: the problem is not with reading:
`git config -f "//hyprostr/dfs/groups/developers/settings/gitconfig.txt" --list` it works fine. But problem with writting. Did you see my video of the problem?

**Comment by kostix** [comment]
I did not see the video (Andrey, posting a video is rather pointless in most cases as it requires context switching, and it's not googleable), but writing to a config file on a share does *not* work form me either:

````
kostix@programmer MINGW32 ~
$ git config -f //tigra/kostix/foo.conf --list
core.whatever=bar

kostix@programmer MINGW32 ~
$ git config -f //tigra/kostix/foo.conf mumbo.jumbo xyzzy
error: could not commit config file //tigra/kostix/foo.conf

kostix@programmer MINGW32 ~
$ git config -f //tigra/kostix/foo.conf --add mumbo.jumbo xyzzy
error: could not commit config file //tigra/kostix/foo.conf
````

I did not see the code (yet) but I have a vague feeling that Git, which had been engeneered by professionals, did not blindly overwrites the file with new content but creates another one in the same directory, `fsync()`s the write then calls `rename(new_name, orig_name)` (and then possibly `fsync()`s the directory entry).

May be this atomic rename is broken in GfW because it does not use [`ReplaceFile`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365512) or calls it with the `lpBackupFileName` set to `NULL` which appears to make a big difference in certain cases.

In either case all this are just my wild guesses, unfortunately.

**Comment by Andrey-Bushman** [comment]
It can be used as workaround of the bug:
`mv "//hyprostr/dfs/groups/developers/settings/gitconfig.txt" "./gitconfig.txt"`
`git config -f "./gitconfig.txt" http.proxy http://@proxy2:8080`
`mv "./gitconfig.txt" "//hyprostr/dfs/groups/developers/settings/gitconfig.txt"`
It is not a nice decission, but it works.

**Comment by drizzd** [comment]
This seems to be the code in question:
https://github.com/git-for-windows/git/blob/master/lockfile.c#L290

We call fclose immediately followed by MoveFileExW. Maybe this is too fast for Windows shares?
https://github.com/git-for-windows/git/blob/master/lockfile.c#L265
https://github.com/git-for-windows/git/blob/master/compat/mingw.c#L1862

**Comment by Andrey-Bushman** [comment]
>Maybe this is too fast for Windows shares?

I don't know, I am not expert in C. To get the expected result is important for me. My previous message does it. I read the `git help config` about the `alias.*`. It is not possible to write these three commands through an aliace. Am I right? I hope I am mistaken, still and it is possible. If I am mistaken, then how can I do it?

**Comment by Andrey-Bushman** [comment]
I solved my problem with alias here: http://stackoverflow.com/questions/31447193/can-i-use-a-set-of-commands-external-and-internal-for-a-git-alias

**Comment by dscho** [comment]
@Andrey-Bushman couldn't you just have pasted the relevant parts in this ticket, to make things easier for everybody else?

**Comment by Andrey-Bushman** [comment]
> couldn't you just have pasted the relevant parts in this ticket, to make things easier for everybody else?

Ok. I wrote such script:

`# edit_config.sh`
`# © Andrey Bushman, 2015`
`# This is a workaround of the problem described here:`
`# https://github.com/git-for-windows/git/issues/241`
`# Arguments:`
`# $1 - target config file`
`# $2 - parameter full name`
`# $3 - parameter value`
`random_file_name=$RANDOM`
`mv $1 $random_file_name`
`git config -f $random_file_name $2 $3`
`mv $random_file_name $1`

The script file I located in the network also. Now it can be launched either through Git Bash directly:

`sh "//hyprostr/dfs/groups/developers/settings/edit_config.sh" "//hyprostr/dfs/groups/developers/settings/gitconfig.txt" "http.proxy" "http://@proxy2:8080"`

or through the alias

`git config --global alias.editconfig '!sh "//hyprostr/dfs/groups/developers/settings/edit_config.sh"'`
I can launch this:

`git editconfig "//hyprostr/dfs/groups/developers/settings/gitconfig.txt" "http.proxy" "http://@proxy2:8080"`
It works fine for me.

**Comment by drizzd** [comment]
I cannot reproduce. Works perfectly for me. :-(

**Comment by Andrey-Bushman** [comment]
Did you see my video? I showed, what **sometimes** it works for me. The keyword is **sometimes**...

**Comment by dscho** [comment]
@Andrey-Bushman A video is probably not the most appropriate way to relate your problem. And if it works **sometimes** you might want to try to figure out **what** are the circumstances when it works vs when it does not, and report back the findings.

**Comment by Andrey-Bushman** [comment]
>A video is probably not the most appropriate way to relate your problem. And if it works sometimes you might want to try to figure out what are the circumstances when it works vs when it does not, and report back the findings.

@dscho, I didn't reveal regularity of this behavior. Also, I haven't time for this spending. Video shows a problem for people, whom this problem isn't reproduced. It shows both variants of working. Also I showed a workaround. I found a problem. I reported about it. I even recorded video. I showed a workaround. I won't persuade developers to correct it. If on your computer the problem isn't reproduced and therefore you consider that it doesn't exist,  then let will be such final of the bug. 

> I cannot reproduce. Works perfectly for me. :-(

It successfully worked for me few times only.
